scrapy/utils/curl.py

```diff
@@ -0,0 +1,80 @@
+import argparse
+from shlex import split
+
+from six.moves.http_cookies import SimpleCookie
+from six.moves.urllib.parse import urlparse, urlunparse
+from six import string_types, iteritems
+from w3lib.http import basic_auth_header
+
+
+class CurlParser(argparse.ArgumentParser):
+    def error(self, message):
+        error_msg = 'There was an error parsing the curl command: {}'.format(message)
+        raise ValueError(error_msg)
+
+
+curl_parser = CurlParser()
+curl_parser.add_argument('url')
+curl_parser.add_argument('-H', '--header', dest='headers', action='append')
+curl_parser.add_argument('-X', '--request', dest='method', default='get')
+curl_parser.add_argument('-d', '--data', dest='data')
+curl_parser.add_argument('-u', '--user', dest='auth')
+curl_parser.add_argument('--compressed', action='store_true')
+curl_parser.add_argument('-s', '--silent', action='store_true')
+
+
+def curl_to_request_kwargs(curl_args):
+    """Convert a curl command syntax to Request kwargs
+
+    :param curl_args: string containing the curl command
+    :return: dictionary of Request kwargs
+    """
+
+    if isinstance(curl_args, string_types):
+        curl_args = split(curl_args)
+
+    if curl_args[0] != 'curl':
+        raise ValueError('A curl command must start with "curl"')
+
+    parsed_args, argv = curl_parser.parse_known_args(curl_args[1:])
+    if argv:
+        msg = 'Unrecognized arguments: %s'
+        raise ValueError(msg % (argv,))
```

Comment:
I'm not sure about this question. I think the way is implemented (accepting some arguments and ignore them but raising those arguments that are not recognized) is the best as if someone uses this method assuming it will generate exactly the same request than the curl command and we just ignore some arguments without raising that could fail without realizing it. However I understand that this could make this code more difficult to mantain.

What do you think? @kmike @Gallaecio

Reply:
(Of course, if we decide to continue like this I should add a lot of parameters to this code).

Reply:
@Gallaecio  What do you thing about adding something like "Note that it has limited support as not all the cURL arguments have sense in Scrapy context" in docstrings?

Reply:
I think there are 2 use cases:

1) make Request similar to a curl command, on a best-effort basis, hoping that it would be enough;
2) make Request exactly as close as possible to curl.

Our main use case is (1), where user tries to reproduce the response from a working curl command, copy-pasting it e.g. from a browser. It means that we shouldn't raise an exception by default if not all options are supported - but having a warning would be nice, as well as a message in a docstring.

The problem with a warning is that we detect that something is not supported, but don't return this information to an end user, so users can't raise an error themselves if they want - so (2) is not supported if we only have a warning. I think this can be solved in several ways, e.g.:

1) have options parsing in a separate function, and return parsed options + leftovers, so that users can check if there are any leftovers, and act on it. 
2) have a flag (ignore_unknown_options=True), similar to request's raise_for_status, which allows user to turn a warning to an exception.